addons/test_mail/tests/test_mail_composer.py

```diff
@@ -1852,6 +1859,79 @@ def test_mail_composer_delete(self):
+            'partner_to': '',
+        })
+
+        # Guarantee points of variation for test_report_3
+        self.test_records[0].write({'name': 'A', 'count': 1})
+        self.test_records[1].write({'name': 'B', 'count': 2})
+
+        attachment = self.env['ir.attachment'].create({
+            'datas': base64.b64encode(b'ExtraData'),
+            'mimetype': 'text/plain',
+            'name': 'SomeCommonAttachment.txt',
+            'res_model': False,
+            'res_id': False,
+        })
+
+        def _instanciate_composer():
+            composer_form = Form(self.env['mail.compose.message'].with_context(
+                self._get_web_context(self.test_records, add_web=True,
+                                      default_template_id=self.template.id)
+            ))
+            return composer_form.save()
+
+        base = {
+            'attachment_ids': [Command.clear()],
+            'body_html': '<p>Common Body</p>',
+            'report_template_ids': [Command.clear()],
+            'subject': 'Common Subject',
+        }
+        same_attachments = {'attachment_ids': attachment.ids}
+        diff_body = {'body_html': '<p><t t-esc="object.name"></t></p>'}
+        # regardless of whether they have different bodies or not, they are considered duplicates
+        diff_attachment_same_content = {'report_template_ids': [self.test_report_2.id]}
+        diff_attachment_diff_content = {'report_template_ids': [self.test_report_3.id]}
+        diff_subject = {'subject': '{{ object.name }}'}
+        # expect all sent
+        for changes in [[diff_body], [diff_subject],
+                        [diff_attachment_same_content], [diff_attachment_same_content],
+                        [diff_attachment_same_content, same_attachments],
+                        [diff_body, diff_attachment_diff_content, diff_subject]]:
+            full_diff = dict(base)
+            for change in changes:
+                full_diff.update(change)
+            self.template.write(full_diff)
+            composer = _instanciate_composer()
+            with self.mock_mail_gateway(mail_unlink_sent=False), self.mock_mail_app():
+                composer._action_send_mail()
+            self.assertEqual(len(self._mails), 2, f'Should have sent all email.\nvalues: {full_diff}')
+
+        # expect duplicates
+        for with_composer_attachment, changes in zip([False, True], [[], [same_attachments]] * 2):
+            full_diff = dict(base)
+            for change in changes:
+                full_diff.update(change)
+            self.template.write(full_diff)
+            composer = _instanciate_composer()
+            if with_composer_attachment:
+                composer.attachment_ids = attachment
+            with self.mock_mail_gateway(mail_unlink_sent=False), self.mock_mail_app():
+                composer._action_send_mail()
+            self.assertEqual(len(self._mails), 1, 'Should have sent 1 email, and skipped a duplicate.\n'
```

Comment:
Could be great to use assertMail with state, to check we effectively have one email sent, and one email canceled, to be sure we log them as intended.

Reply:
@tde-banana-odoo Done, had to give `common.py` utils a touch up too so they would behave properly. I think it should be ok, other tests seem unaffected.